### PR TITLE
[WIP] Plutus core/cgp 205

### PR DESCRIPTION
**Pull request by vmchale**
This contains type/kind synthesis.

### CODE REVIEW

**Comment by mchakravarty** [comment]
Oh, and why are the tests failing?

File: language-plutus-core/src/Language/PlutusCore/Lexer/Type.hs [hunk 1]
```diff
@@ -21,10 +21,10 @@ import           Numeric                            (showHex)
 import           PlutusPrelude
 
 -- | A builtin type
-data TypeBuiltin = TyByteString
+data TypeBuiltin = TyByteString -- FIXME these should take integer/naturals
```
[the remark this answers is not in the document]
Comment (vmchale) [reply language-plutus-core/src/Language/PlutusCore/Lexer/Type.hs]: The bytestring type requires a size. We do not do anything about that here.

**Comment by vmchale** [comment]
I think a good first stab at testing would be golden tests. This could ensure our error messages were reasonably well-formed as well.

I'm not sure what's up with the CI. Does it work on `master`?

**Comment by mchakravarty** [comment]
Ok, let's do golden tests for now. The problem with that is that it is quite tricky to find a somewhat comprehensive set of tests. We should at least cover everything single AST construction in one of the tests.

File: language-plutus-core/src/Language/PlutusCore/Lexer/Type.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Lexer/Type.hs]: > More generally, we need to think about testing for the type checker. What do you think? How should be approach it?

We could generate well-typed terms and type check them. This is not an easy task and there is a lot of literature (for a couple of examples: [an entire thesis](https://gupea.ub.gu.se/bitstream/2077/36981/1/gupea_2077_36981_1.pdf), a [paper](https://pdfs.semanticscholar.org/782d/823cb0499a969f3cd0ca7aaba0bcf892fba3.pdf)), but it's [non-trivial to generate interesting terms](http://okmij.org/ftp/Haskell/AlgorithmsH1.html#random-term) even in an untyped setting. Could be useful eventually.

But I think having a lot of golden tests that cover known corner cases is already a very good thing. This is what most production compilers do, I believe.

**Comment by effectfully** [comment]
> More generally, we need to think about testing for the type checker. What do you think? How should be approach it?

We could generate well-typed terms and type check them. This is not an easy task and there is a lot of literature (for a couple of examples: an [entire thesis](https://gupea.ub.gu.se/bitstream/2077/36981/1/gupea_2077_36981_1.pdf), a [paper](https://pdfs.semanticscholar.org/782d/823cb0499a969f3cd0ca7aaba0bcf892fba3.pdf)), but it's [non-trivial to generate interesting terms](http://okmij.org/ftp/Haskell/AlgorithmsH1.html#random-term) even in an untyped setting. Could be useful at some point.

But I think having a lot of golden tests that cover known corner cases is already a very good thing. This is what most production compilers do, I believe.

**Comment by mchakravarty** [comment]
@effectfully 
> We could generate well-typed terms and type check them. This is not an easy task and there is a lot of literature (for a couple of examples: an entire thesis, a paper), but it's non-trivial to generate interesting terms even in an untyped setting. Could be useful at some point.

Yes, I think, this would be very useful. As you are saying, it will require some effort, but I think, it would make for a good way to test the type checker. We'll probably want to finish the main components of this library first, but then we should have a look at it.

**Comment by vmchale** [comment]
@effectfully for now I'll focus on having comprehensive coverage with golden tests, but I agree with you: generating well-typed programs will catch many errors.
